crates/accelerate/src/sabre_swap/mod.rs

```diff
@@ -390,46 +496,119 @@ fn update_route(
     required_predecessors: &mut [u32],
-) {
+    node_block_results: &mut HashMap<usize, Vec<BlockResult>>,
+    route_block_dag: &F,
+) where
+    F: Fn(&SabreDAG, NLayout) -> (SabreResult, NLayout),
+{
     let mut to_visit = to_visit.to_vec();
     let mut i = 0;
     // Iterate through `to_visit`, except we often push new nodes onto the end of it.
     while i < to_visit.len() {
         let node = to_visit[i];
         i += 1;
         let (py_node, qubits) = &dag.dag[node];
-        match qubits[..] {
-            [a, b]
-                if !coupling.contains_edge(
-                    NodeIndex::new(layout.logic_to_phys[a]),
-                    NodeIndex::new(layout.logic_to_phys[b]),
-                ) =>
-            {
-                front_layer.insert(node, [a, b]);
+
+        match dag.node_blocks.get(py_node) {
+            Some(blocks) => {
+                // Control flow op. Route all blocks for current layout.
+                let mut block_results: Vec<BlockResult> = Vec::with_capacity(blocks.len());
+                for inner_dag in blocks {
+                    let (inner_dag_routed, inner_final_layout) =
+                        route_block_dag(inner_dag, layout.copy());
+
+                    // For now, we always append a swap circuit that gets the inner block
+                    // back to the parent's layout.
+                    let swap_epilogue =
+                        gen_swap_epilogue(coupling, inner_final_layout, layout, seed);
+                    let block_result = BlockResult {
+                        result: inner_dag_routed,
+                        swap_epilogue,
+                    };
+                    block_results.push(block_result);
+                }
+                node_block_results.insert_unique_unchecked(*py_node, block_results);
             }
-            _ => {
-                gate_order.push(*py_node);
-                for edge in dag.dag.edges_directed(node, Direction::Outgoing) {
-                    let successor_node = edge.target();
-                    let successor_index = successor_node.index();
-                    required_predecessors[successor_index] -= 1;
-                    if required_predecessors[successor_index] == 0 {
-                        to_visit.push(successor_node);
-                    }
+            None => match qubits[..] {
+                // A gate op whose connectivity must match the device to be
+                // placed in the gate order.
+                [a, b]
+                    if !coupling.contains_edge(
+                        NodeIndex::new(layout.logic_to_phys[a]),
+                        NodeIndex::new(layout.logic_to_phys[b]),
+                    ) =>
+                {
+                    // 2Q op that cannot be placed. Add it to the front layer
+                    // and move on.
+                    front_layer.insert(node, [a, b]);
+                    continue;
                 }
+                _ => {}
```

Comment:
Huh, I guess `_ => {}` _is_ shorter than `_ => (),` haha.  For some reason, I've always the latter.